From 77ccd47f63e53bb08db9e52b7766ef2f6bf20dc0 Mon Sep 17 00:00:00 2001
From: Jitendra Pandey <jitendra@apache.org>
Date: Fri, 7 Jul 2017 12:45:37 -0700
Subject: [PATCH 2690/2848] HADOOP-10829. Iteration on
 CredentialProviderFactory.serviceLoader is
 thread-unsafe. Contributed by Benoy Antony and
 Rakesh R.

(cherry picked from commit f1efa14fc676641fa15c11d3147e3ad948b084e9)

Change-Id: I143df30b4b209580753eca29c7aba7ac24e90aa8
Ref: CDH-61947
---
 .../security/alias/CredentialProviderFactory.java  |   10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialProviderFactory.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialProviderFactory.java
index d1e3eb5..1b2ac41 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialProviderFactory.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialProviderFactory.java
@@ -22,6 +22,7 @@
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
 import java.util.ServiceLoader;
 
@@ -49,6 +50,15 @@ public abstract CredentialProvider createProvider(URI providerName,
       ServiceLoader.load(CredentialProviderFactory.class,
           CredentialProviderFactory.class.getClassLoader());
 
+  // Iterate through the serviceLoader to avoid lazy loading.
+  // Lazy loading would require synchronization in concurrent use cases.
+  static {
+    Iterator<CredentialProviderFactory> iterServices = serviceLoader.iterator();
+    while (iterServices.hasNext()) {
+      iterServices.next();
+    }
+  }
+
   public static List<CredentialProvider> getProviders(Configuration conf
                                                ) throws IOException {
     List<CredentialProvider> result = new ArrayList<CredentialProvider>();
-- 
1.7.9.5

